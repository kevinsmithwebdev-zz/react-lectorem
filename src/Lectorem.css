.Lectorem {
  align-items: center;
  display: flex;
  flex-direction: column;
  margin: 5px;
  padding: 5px;
  background-color: #fff;
  width: 800px;
}

audio {
  filter: invert(20%);
  width: 400px;
  height: 50px;
}

.Lectorem .playerContainer {
  padding: 5px;
}

.Lectorem-paragraph-header {
  text-align: center;
  margin: 5px;
  padding: 5px;
}

.Lectorem .paragraph {
  text-align: left;
  margin: 5px 10px;;
  padding: 5px;
}

.Lectorem h2 {
  margin: 5px;
  padding: 0;
}

.Lectorem h3 {
  margin: 5px;
  padding: 0;
}
