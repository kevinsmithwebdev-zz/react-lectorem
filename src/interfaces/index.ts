export interface LectInterface {
  start?: number,
  end?: number,
  text: string,
  translation?: string,
  explanation?: string,
}

export interface SentenceInterface {
  lects: LectInterface[],
  sentenceTranslation?: string,
}

export interface HeaderInterface {
  imagePath?: string,
}

export interface ParagraphInterface {
  header?: HeaderInterface,
  sentences: SentenceInterface[],
}

export interface StoryDataInterface {
  audioPath: string,
  title?: string,
  subtitle?: string,
  paragraphs: ParagraphInterface[],
}

export interface ConfigurationInterface {
  shouldShowTranslation?: boolean,
  isTranslationBelow?: boolean,
  isBySentence?: boolean,
}