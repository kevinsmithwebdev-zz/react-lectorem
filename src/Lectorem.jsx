import React, { useState } from 'react';
import Player from './Player/Player';
import './Lectorem.css';
import Modal from './Modal/Modal';
import Texts from './Texts/Texts';

export const createShowModal = setModalData =>
  lect => (lect.translation || lect.explanation) && setModalData(lect);

export const createHideModal = setModalData => () => setModalData(null);

const Lectorem = ({ data, configuration }) => {
  const [modalData, setModalData] = useState(null);
  const [readTime, setReadTime] = useState(null);

  if (!data) {
    return <p>Loading...</p>;
  }

  const showModal = createShowModal(setModalData);
  const hideModal = createHideModal(setModalData);

  const {
    title, subtitle, audioPath, paragraphs,
  } = data;

  return (
    <div className='Lectorem'>
      { !!title && <h2>{title} </h2> }
      { !!subtitle && <h3>{subtitle} </h3> }
      <div className='playerContainer'>
        <Player audioPath={ audioPath } onListen={ setReadTime } />
      </div>
      <div className='lectsContainer'>
        <Texts
          configuration={ configuration }
          paragraphs={ paragraphs }
          readTime={ readTime }
          showModal={ showModal }
        />
      </div>

      <Modal data={ modalData } hideModal={ hideModal } />
    </div>
  );
}

export default Lectorem;
