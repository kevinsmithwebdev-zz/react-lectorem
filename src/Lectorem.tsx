import React, { useState } from 'react';
import Player from './Player/Player';
import classes from './Lectorem.module.css';
import Modal from './Modal/Modal';
import Texts from './Texts/Texts';
import { StoryDataInterface, ConfigurationInterface, LectInterface } from './interfaces/index';

interface ComponentPropsInterface {
  data: StoryDataInterface,
  configuration?: ConfigurationInterface,
}

export const createShowModal =
  (setModalData: Function): Function =>
    (lect: LectInterface): void => (lect.translation || lect.explanation) && setModalData(lect);

export const createHideModal =
  (setModalData: Function): Function =>
    (): void => setModalData(null);

const Lectorem: React.FC<ComponentPropsInterface> = ({ data, configuration }) => {
  const [modalData, setModalData] = useState(null);
  const [readTime, setReadTime] = useState(null);

  if (!data) {
    return <p>Loading...</p>;
  }

  const showModal = createShowModal(setModalData);
  const hideModal = createHideModal(setModalData);

  const {
    title, subtitle, audioPath, paragraphs,
  } = data;

  return (
    <div className={ classes.Lectorem }>
      { !!title && <h2>{title} </h2> }
      { !!subtitle && <h3>{subtitle} </h3> }
      <div className={ classes.playerContainer }>
        <Player audioPath={ audioPath } onListen={ setReadTime } />
      </div>
      <div className={ classes.lectsContainer }>
        <Texts
          configuration={ configuration }
          paragraphs={ paragraphs }
          readTime={ readTime }
          showModal={ showModal }
        />
      </div>

      <Modal data={ modalData } hideModal={ hideModal } />
    </div>
  );
}

export default Lectorem;
