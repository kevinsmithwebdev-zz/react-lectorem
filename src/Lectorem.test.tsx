import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Lectorem, { createShowModal, createHideModal } from './Lectorem';
import storyData from './fixtureData/story.data';
import { StoryDataInterface } from './interfaces/index';

describe('Lectorem', () => {
  describe('the main component render', () => {
    const renderer = ShallowRenderer.createRenderer();

    describe('with data', () => {
      describe('with no data', () => {
        const props = {
          data: undefined as StoryDataInterface,
        };
        it('should render without crashing', () => {
          renderer.render(<Lectorem {...props} />);
          expect(renderer.getRenderOutput()).toMatchSnapshot();
        });
      });

      describe('with all data', () => {
        const props = {
          data: storyData,
          configuration: { isTranslationShown: false, isTranslationBelow: false, isBySentence: false },
        };
        it('should render without crashing', () => {
          renderer.render(<Lectorem {...props} />);
          expect(renderer.getRenderOutput()).toMatchSnapshot();
        });
      });

      describe('with all data except title and subtitle', () => {
        const props = {
          data: {
            ...storyData,
            title: undefined as string,
            subtitle: undefined as string,
          },
          configuration: { isTranslationShown: false, isTranslationBelow: false, isBySentence: false },
        };
        it('should render without crashing', () => {
          renderer.render(<Lectorem {...props} />);
          expect(renderer.getRenderOutput()).toMatchSnapshot();
        });
      });
    });
  });
  describe('createShowModal', () => {
    const setModalData = jest.fn();
    const showModal = createShowModal(setModalData);
    describe('the function it returns', () => {
      describe('when lect has translation but no explanation', () => {
        const lect = { translation: 'translation123' };
        beforeAll(() => {
          setModalData.mockReset();
          showModal(lect);
        });
        it('should call setModalData with correct lect', () => {
          expect(setModalData).toBeCalledWith(lect);
        });
      });

      describe('when lect no translation but has explanation', () => {
        const lect = { explanation: 'explanation123' };
        beforeAll(() => {
          setModalData.mockReset();
          showModal(lect);
        });
        it('should call setModalData with correct lect', () => {
          expect(setModalData).toBeCalledWith(lect);
        });
      });

      describe('when lect has neither translation nor explanation', () => {
        const lect = { other: 'data' };
        beforeAll(() => {
          setModalData.mockReset();
          showModal(lect);
        });
        it('should call not call setModalData ', () => {
          expect(setModalData).not.toBeCalled();
        });
      });
    });
  });

  describe('createHideModal', () => {
    const setModalData = jest.fn();
    const hideModal = createHideModal(setModalData);
    describe('the function it returns', () => {
      beforeAll(() => {
        setModalData.mockReset();
        hideModal();
      });
      it('should call setModalData with null', () => {
        expect(setModalData).toBeCalledWith(null);
      });
    });
  });
});
