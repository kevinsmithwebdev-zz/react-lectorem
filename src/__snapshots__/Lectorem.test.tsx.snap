// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lectorem the main component render with data with all data except title and subtitle should render without crashing 1`] = `
<div
  className="Lectorem"
>
  <div
    className="playerContainer"
  >
    <Player
      audioPath="./frere.mp3"
      onListen={[Function]}
    />
  </div>
  <div
    className="lectsContainer"
  >
    <Texts
      configuration={
        Object {
          "isBySentence": false,
          "isTranslationBelow": false,
          "isTranslationShown": false,
        }
      }
      paragraphs={
        Array [
          Object {
            "header": Object {
              "imagePath": "./frere.png",
            },
            "sentences": Array [
              Object {
                "lects": Array [
                  Object {
                    "end": 5.511,
                    "explanation": "\\"Brother\\" in this sense means the title of a monk. \\"Frère\\" is also where we get our word \\"friar\\"",
                    "start": 4.424,
                    "text": "Frère",
                    "translation": "Brother",
                  },
                  Object {
                    "end": 6.533,
                    "explanation": "In the English version of the song this is rendered as \\"John\\" but a more accurate translation would be \\"James\\" or \\"Jacob\\", from the Latin \\"Iacobus\\".",
                    "start": 5.51,
                    "text": "Jacques",
                    "translation": "James",
                  },
                  Object {
                    "text": ".",
                  },
                ],
                "sentenceTranslation": "Brother James.",
              },
              Object {
                "lects": Array [
                  Object {
                    "end": 7.683,
                    "explanation": "\\"Brother\\" in this sense means the title of a monk. \\"Frère\\" is also where we get our word \\"friar\\"",
                    "start": 6.618,
                    "text": "Frère",
                    "translation": "Brother",
                  },
                  Object {
                    "end": 8.79,
                    "explanation": "In the English version of the song this is rendered as \\"John\\" but a more accurate translation would be \\"James\\" or \\"Jacob\\", from the Latin \\"Iacobus\\".",
                    "start": 7.68,
                    "text": "Jacques",
                    "translation": "James",
                  },
                  Object {
                    "text": ".",
                  },
                ],
                "sentenceTranslation": "Brother James.",
              },
            ],
          },
          Object {
            "sentences": Array [
              Object {
                "lects": Array [
                  Object {
                    "end": 10.857,
                    "start": 8.831,
                    "text": "Dormez-vous?",
                    "translation": "Are you sleeping?",
                  },
                ],
                "sentenceTranslation": "Are you sleeping?",
              },
              Object {
                "lects": Array [
                  Object {
                    "end": 12.71,
                    "start": 11.025,
                    "text": "Dormez-vous?",
                    "translation": "Are you sleeping?",
                  },
                ],
                "sentenceTranslation": "Are you sleeping?",
              },
            ],
          },
          Object {
            "sentences": Array [
              Object {
                "lects": Array [
                  Object {
                    "end": 13.706,
                    "start": 13.21,
                    "text": "Sonnez",
                    "translation": "Ring",
                  },
                  Object {
                    "end": 15.399,
                    "explanation": "These were the bells to wake the monks for the \\"matins\\", the first chants in the day, very early in the morning.",
                    "start": 13.704,
                    "text": "les matines",
                    "translation": "the morning bells",
                  },
                  Object {
                    "text": "!",
                  },
                ],
                "sentenceTranslation": "Ring the morning bells!",
              },
              Object {
                "lects": Array [
                  Object {
                    "end": 15.924,
                    "start": 15.39,
                    "text": "Sonnez",
                    "translation": "Ring",
                  },
                  Object {
                    "end": 17.572,
                    "explanation": "These were the bells to wake the monks for the \\"matins\\", the first chants in the day, very early in the morning.",
                    "start": 15.923,
                    "text": "les matines",
                    "translation": "the morning bells",
                  },
                  Object {
                    "text": "!",
                  },
                ],
                "sentenceTranslation": "Ring the morning bells!",
              },
            ],
          },
          Object {
            "sentences": Array [
              Object {
                "lects": Array [
                  Object {
                    "end": 19.319,
                    "explanation": "An onomatopoeia for bells, like our \\"ding dong\\".",
                    "start": 17.57,
                    "text": "Ding, dang, dong.",
                  },
                ],
                "sentenceTranslation": "Ding, dang, dong.",
              },
              Object {
                "lects": Array [
                  Object {
                    "end": 21.632,
                    "explanation": "An onomatopoeia for bells, like our \\"ding dong\\".",
                    "start": 19.784,
                    "text": "Ding, dang, dong.",
                  },
                ],
                "sentenceTranslation": "Ding, dang, dong.",
              },
            ],
          },
        ]
      }
      readTime={null}
      showModal={[Function]}
    />
  </div>
  <Modal
    data={null}
    hideModal={[Function]}
  />
</div>
`;

exports[`Lectorem the main component render with data with all data should render without crashing 1`] = `
<div
  className="Lectorem"
>
  <h2>
    Frère Jacques
     
  </h2>
  <h3>
    (Are You Sleeping)
     
  </h3>
  <div
    className="playerContainer"
  >
    <Player
      audioPath="./frere.mp3"
      onListen={[Function]}
    />
  </div>
  <div
    className="lectsContainer"
  >
    <Texts
      configuration={
        Object {
          "isBySentence": false,
          "isTranslationBelow": false,
          "isTranslationShown": false,
        }
      }
      paragraphs={
        Array [
          Object {
            "header": Object {
              "imagePath": "./frere.png",
            },
            "sentences": Array [
              Object {
                "lects": Array [
                  Object {
                    "end": 5.511,
                    "explanation": "\\"Brother\\" in this sense means the title of a monk. \\"Frère\\" is also where we get our word \\"friar\\"",
                    "start": 4.424,
                    "text": "Frère",
                    "translation": "Brother",
                  },
                  Object {
                    "end": 6.533,
                    "explanation": "In the English version of the song this is rendered as \\"John\\" but a more accurate translation would be \\"James\\" or \\"Jacob\\", from the Latin \\"Iacobus\\".",
                    "start": 5.51,
                    "text": "Jacques",
                    "translation": "James",
                  },
                  Object {
                    "text": ".",
                  },
                ],
                "sentenceTranslation": "Brother James.",
              },
              Object {
                "lects": Array [
                  Object {
                    "end": 7.683,
                    "explanation": "\\"Brother\\" in this sense means the title of a monk. \\"Frère\\" is also where we get our word \\"friar\\"",
                    "start": 6.618,
                    "text": "Frère",
                    "translation": "Brother",
                  },
                  Object {
                    "end": 8.79,
                    "explanation": "In the English version of the song this is rendered as \\"John\\" but a more accurate translation would be \\"James\\" or \\"Jacob\\", from the Latin \\"Iacobus\\".",
                    "start": 7.68,
                    "text": "Jacques",
                    "translation": "James",
                  },
                  Object {
                    "text": ".",
                  },
                ],
                "sentenceTranslation": "Brother James.",
              },
            ],
          },
          Object {
            "sentences": Array [
              Object {
                "lects": Array [
                  Object {
                    "end": 10.857,
                    "start": 8.831,
                    "text": "Dormez-vous?",
                    "translation": "Are you sleeping?",
                  },
                ],
                "sentenceTranslation": "Are you sleeping?",
              },
              Object {
                "lects": Array [
                  Object {
                    "end": 12.71,
                    "start": 11.025,
                    "text": "Dormez-vous?",
                    "translation": "Are you sleeping?",
                  },
                ],
                "sentenceTranslation": "Are you sleeping?",
              },
            ],
          },
          Object {
            "sentences": Array [
              Object {
                "lects": Array [
                  Object {
                    "end": 13.706,
                    "start": 13.21,
                    "text": "Sonnez",
                    "translation": "Ring",
                  },
                  Object {
                    "end": 15.399,
                    "explanation": "These were the bells to wake the monks for the \\"matins\\", the first chants in the day, very early in the morning.",
                    "start": 13.704,
                    "text": "les matines",
                    "translation": "the morning bells",
                  },
                  Object {
                    "text": "!",
                  },
                ],
                "sentenceTranslation": "Ring the morning bells!",
              },
              Object {
                "lects": Array [
                  Object {
                    "end": 15.924,
                    "start": 15.39,
                    "text": "Sonnez",
                    "translation": "Ring",
                  },
                  Object {
                    "end": 17.572,
                    "explanation": "These were the bells to wake the monks for the \\"matins\\", the first chants in the day, very early in the morning.",
                    "start": 15.923,
                    "text": "les matines",
                    "translation": "the morning bells",
                  },
                  Object {
                    "text": "!",
                  },
                ],
                "sentenceTranslation": "Ring the morning bells!",
              },
            ],
          },
          Object {
            "sentences": Array [
              Object {
                "lects": Array [
                  Object {
                    "end": 19.319,
                    "explanation": "An onomatopoeia for bells, like our \\"ding dong\\".",
                    "start": 17.57,
                    "text": "Ding, dang, dong.",
                  },
                ],
                "sentenceTranslation": "Ding, dang, dong.",
              },
              Object {
                "lects": Array [
                  Object {
                    "end": 21.632,
                    "explanation": "An onomatopoeia for bells, like our \\"ding dong\\".",
                    "start": 19.784,
                    "text": "Ding, dang, dong.",
                  },
                ],
                "sentenceTranslation": "Ding, dang, dong.",
              },
            ],
          },
        ]
      }
      readTime={null}
      showModal={[Function]}
    />
  </div>
  <Modal
    data={null}
    hideModal={[Function]}
  />
</div>
`;

exports[`Lectorem the main component render with data with no data should render without crashing 1`] = `
<p>
  Loading...
</p>
`;
